addons/stock_account/wizard/stock_valuation_layer_revaluation.py

```diff
@@ -0,0 +1,146 @@
+    _name = 'stock.valuation.layer.revaluation'
+    _description = "Wizard model to reavaluate a stock inventory for a product"
+    _check_company_auto = True
+
+    @api.model
+    def default_get(self, default_fields):
+        res = super().default_get(default_fields)
+        if res.get('product_id'):
+            product = self.env['product.product'].browse(res['product_id'])
+            accounts = product.product_tmpl_id.get_product_accounts()
+            if product.categ_id.property_cost_method == 'standard':
+                raise UserError(_("You cannot revalue a product with a standard cost method."))
+            if product.quantity_svl <= 0:
+                raise UserError(_("You cannot revalue a product with an empty or negative stock."))
+            if 'account_journal_id' not in res and 'account_journal_id' in default_fields:
+                res['account_journal_id'] = accounts['stock_journal'].id
+        return res
+
+    company_id = fields.Many2one('res.company', "Company", readonly=True, required=True)
+    currency_id = fields.Many2one('res.currency', "Currency", related='company_id.currency_id', required=True)
+
+    product_id = fields.Many2one('product.product', "Related product", required=True, check_company=True)
+    property_valuation = fields.Selection(related='product_id.categ_id.property_valuation')
+    product_uom_name = fields.Char("Unit of Measure", related='product_id.uom_id.name')
+    current_value_svl = fields.Float("Current Value", related="product_id.value_svl")
+    current_quantity_svl = fields.Float("Current Quantity", related="product_id.quantity_svl")
+
+    added_value = fields.Monetary("Added value", required=True)
+    new_value = fields.Monetary("New value", compute='_compute_new_value')
+    new_value_by_qty = fields.Monetary("New value by quantity", compute='_compute_new_value')
+    reason = fields.Char("Reason", help="Reason of the revaluation")
+
+    account_journal_id = fields.Many2one('account.journal', "Journal", required=True, check_company=True)
+    account_id = fields.Many2one('account.account', "Counterpart Account", domain=[('deprecated', '=', False)], required=True, check_company=True)
+    date = fields.Date("Accounting Date")
+
+    @api.depends('current_value_svl', 'current_quantity_svl', 'added_value')
+    def _compute_new_value(self):
+        for reval in self:
+            reval.new_value = reval.current_value_svl + reval.added_value
+            if not self.company_id.currency_id.is_zero(reval.current_quantity_svl):
+                reval.new_value_by_qty = reval.new_value / reval.current_quantity_svl
+            else:
+                reval.new_value_by_qty = 0.0
+
+    def action_validate_revaluation(self):
+        """ Revaluate the stock for `self.product_id` in `self.company_id`.
+
+        - Change the stardard price with the new valuation by product unit.
+        - Create a manual stock valuation layer with the `added_value` of `self`.
```

Comment:
- Distribute the added value on the `remaining_value` of the layers still in stock.